ocaml/typing/printtyp.ml

```diff
@@ -1261,27 +1261,44 @@ let out_jkind_of_user_jkind (jkind : Jane_syntax.Jkind.annotation) =
           modes.txt
       in
       Ojkind_user_mod (base, modes)
-    | With _ | Kind_of _ | Arrow _ -> failwith "XXX unimplemented jkind syntax"
+    | Arrow (args, result) -> Ojkind_user_arrow {
+        args = List.map out_jkind_user_of_user_jkind args;
+        result = out_jkind_user_of_user_jkind result }
+    | With _ | Kind_of _ -> failwith "XXX unimplemented jkind syntax"
   in
   Ojkind_user (out_jkind_user_of_user_jkind jkind.txt)
 
 let out_jkind_of_const_jkind jkind =
   Ojkind_const (Jkind.Type.Const.to_out_jkind_const jkind)
 
+let rec out_jkind_of_jkind jkind = match Jkind.get jkind with
+  | Type ty -> begin match Jkind.Type.get ty with
+    | Const clay -> out_jkind_of_const_jkind clay
+    | Var v      -> Ojkind_var (Jkind.Type.Sort.Var.name v)
+    end
+  | Arrow { args; result } ->
+    Ojkind_arrow
+      { args = List.map out_jkind_of_jkind args;
+        result = out_jkind_of_jkind result }
+
 (* returns None for [value], according to (C2.1) from
    Note [When to print jkind annotations] *)
 let out_jkind_option_of_jkind jkind =
   match Jkind.get jkind with
-  | Const jkind ->
-    let value_jkind = Jkind.Type.Const.Primitive.value.jkind |> Jkind.Const.of_type_jkind in
-    begin match Jkind.Const.equal jkind value_jkind with
-    | true -> None
-    | false -> Some (out_jkind_of_const_jkind jkind)
+  | Type ty -> begin
+    match Jkind.Type.get ty with
+    | Const jkind -> begin
+      match Jkind.Type.Const.equal jkind Jkind.Type.Const.Primitive.value.jkind with
+      | true -> None
+      | false -> Some (out_jkind_of_const_jkind jkind)
+      end
+    | Var v -> (* This handles (X1). *)
+      if !Clflags.verbose_types
+      then Some (Ojkind_var (Jkind.Type.Sort.Var.name v))
+      else None
     end
-  | Var v -> (* This handles (X1). *)
-    if !Clflags.verbose_types
-    then Some (Ojkind_var (Jkind.Type.Sort.Var.name v))
-    else None
+  (* We ignore the rules above for arrows, which should always be printed *)
```

Comment:
I agree we need to print the arrow, but we should still avoid printing the sort variables unless `!Clflags.verbose_types` is set. I'd suggest just printing `_` when we have a sort variable within an arrow. Probably we should update the syntax to accept that syntax, although we don't need to do that right now.

Reply:
This sounds fine.
Side question: isn't it become a bit inconsistent that in the type jkind case both `value` and a sort variable don't get printed at all (this is irrespective of this change)?